로또 1단계 미션 제출합니다

iimpala opened the pull request.

Jjungs7 commented on src/test/java/lotto/domain/LottoMatcherTest.java:
```diff
@@ -0,0 +1,85 @@
+    }
+}
+
+class lottoNumberArgumentsProvider implements ArgumentsProvider {
```
모든 MatchResult Enum에 대해 테스트하기 위해 custom provider를 만들었습니다. 모든 Enum에 하지 않고 몇 개의 Enum만 테스트해도 됐을까요? 랜덤 숫자는 경계값이 확실해서 그 주변을 테스트하면 될 것 같은데 Enum의 경우에는 모든 값이 다 중요해보여서 모두 테스트하게 되었어요

Jjungs7 commented on src/test/java/lotto/domain/LottoMatcherTest.java:
```diff
@@ -0,0 +1,85 @@
+        );
+    }
+
+    private List<LottoNumber> lottoNumberArgument(int... parameters) {
```
이 메소드는 여러 Test파일에서 사용하고 있어서 분리하고싶은데 test 패키지에 utils패키지를 만들어서 분리해도 될까요?

Jjungs7 commented on src/test/java/lotto/domain/LottoMatcherTest.java:
```diff
@@ -0,0 +1,84 @@
+                )));
+    }
+
+    private List<LottoNumber> customLottoNumbers(int... parameters) {
```
이 메소드는 여러 Test파일에서 사용하고 있어서 분리하고싶은데 test 패키지에 utils패키지를 만들어서 분리해도 될까요?

Hue9010 reviewed (changes requested):
안녕하세요. 이번 리뷰를 맡은 휴입니다!
주말이라 리뷰가 조금 늦었네요. 🙏 
질문에 대한 답변은 피드백에 같이 드렸어요. 
피드백과 함께 확인하시고 궁금하신 사항은 언제든 문의주세요.

Jjungs7 commented on src/main/java/lotto/domain/LottoMatcher.java:
```diff
@@ -0,0 +1,49 @@
+        return MatchResult.valueOf(matchCount, matchBonus);
+    }
+
+    private int calcMatchCount(LottoNumbers lottoTicket) {
+        int[] count = new int[1];
+        LottoNumbers luckyNumbers = winningNumbers.getLuckyNumbers();
+
+        luckyNumbers.delegate(lottoNumbers ->
+                count[0] = (int) lottoNumbers.stream()
+                .filter(lottoTicket::contains)
+                .count());
+
+        return count[0];
+    }
```
consumer가 아무 값도 리턴하지 않는다는 점과 lambda에서 primitive 사용하는데 제한이 있다는 점에서 눈치챘어야했는데 끝까지 꼼수를 쓰게 되었네요. 보내주신 글을 읽으면서 lambda의 scope에 관하여 더 이해할 수 있었고 자바는 변수를 초기화할 때 effectively final로 초기화 한다는 것을 배웠습니다 😆

consumer을 사용한 부분은 궁금한 점을 해소한 다음 수정하고싶어서 질문 드리고자 합니다! 먼저, Consumer를 적극 사용한 이유는 작업을 상위 클래스에서 정하고 이것을 위임한다는 것이 consumer가 작동하는 방식과 일치한다고 생각해서 활용하게 되었어요. 제가 아직 작업을 위임한다는 것을 제대로 이해하지 못해서 그런 것 같습니다 😢  아래 헷갈리는 상황을 적어볼게요!

- 첫 번째 상황: [LottoMatcher.match(LottoTickets lottoTickets)](https://github.com/next-step/java-lotto-kakao/pull/24/files/86b0b9f97f3d528b768c4960518b18db117c7259#diff-68b1175df01ea775d0999fa9ac83597a80d12bce619aab476b6efe69858423c4R17) 에서 `LottoTickets`가 *tickets*을 저장하고 있고 당첨번호와 비교하기 위해서는 *tickets*에 대한 정보가 필요한 상황입니다. 만약 `LottoTickets`에다가 작업을 위임한다고 하면 `LottoTickets.getMatched(WinningNumbers numbers)` 와 같은 메소드를 만들어서 `LottoTickets`가 정보를 반환하게 하는 것이 맞을까요? 그렇게 된다면 아래와 같은 코드가 나올 것으로 예상이 되는데 match() 함수가 하는 일이 너무 없어지는 것이 아닌가 의문이 들어요 ㅠㅠ
```java
// LottoMatcher.java
...
public MatchResults match(LottoTickets lottoTickets) {
    return lottoTickets.getMatched(winningNumbers);
}

// LottoTickets.java
...
public MatchResults getMatched(WinningNumbers numbers) {
    // 당첨 번호와 로또 티켓 뭉치들을 비교하여 몇 개가 당첨됐는지 반환
}
```

- 두 번째 상황: [OutputView.printStatistics(LottoStatistics statistics)](https://github.com/next-step/java-lotto-kakao/pull/24/files/86b0b9f97f3d528b768c4960518b18db117c7259#diff-940d1c44120ea83f932b0ee11c19b877bf63456804f1092acf8a622ed268e26cR29)에서 당첨된 개수를 출력하려면 `MatchResults`에 대한 정보가 모두 필요한 상태입니다. 여기서 `MatchResults`에 작업을 위임하려고 하는데, `MatchResults.constructOutput()` 같은 메소드를 만들어서 필요한 출력 메시지를 구성하려니 view에 관한 로직이 MatchResults에 들어가는 것 같고(또 OutputView는 너무 하는 일이 없어지는 것 같고) `MatchResults.getResults()` getter를 통해서 `EnumMap` 를 전부 가져오자니 내부 데이터를 숨기자는 OOP원칙에 위배되는 것 같아요. 이런 상황에서는 어떻게 하는게 맞을까요??

Hue9010 replied:
### 첫번 째 :
[역할, 책임, 협력](https://velog.io/@ljinsk3/%EC%97%AD%ED%95%A0-%EC%B1%85%EC%9E%84-%ED%98%91%EB%A0%A5), 
[생각하라, 객체지향처럼](https://woowabros.github.io/study/2016/07/07/think_object_oriented.html)를 참고하면 좋을 것 같아요. 구현할 코드가 별로 없는 것(하는 일이 적은 것)은 고려하실 필요가 없습니다. 둘 다 조영호님의 책(객체지향의 사실과 오해, 오브젝트)이 기반인데 가볍게(?) 볼 수 있는 `객체지향의 사실과 오해`를 추천드립니다.

> LottoMatcher: 유저가 가진 로또 묶음과 지난 주 당첨 번호를 비교하여 등수별로 몇 개가 당첨되었는지 판별
 
유저가 가진 로또 묶음을 받아 당첨 번호와 비교하여 결과를 내는 역할을 충분히 담당한다고 볼 수 있습니다. 현실에서는 관리자의 여러 일 중 작업자한테 업무를 배정해주는 일만 대입해서 생각해보면 관리자는 업무만 주고 결과만 받으니 아무 일을 안하는 걸까요? 🤔 협력이라는게 생각보다 별거 없답니다. 😉

### 두번 째 :
넵, 출력메시지를 어떻게 다룰지는 View의 영역입니다. 이 부분에서 DTO로 제공하여 사용하면 View에서 그런 고민을 할 필요가 없습니다. StatisticsResult를 DTO로 만든다면 StatisticsResult의 필드도 모두 DTO로 만들어야 합니다. 그러면 MatchResults의 Map<MatchResult, Integer>를 사용하는 것도 문제가 없죠. 사실 지금은 View가 자바환경이지만 웹 페이지라면 statisticsResult.getResultCountOfSomeMatch같이 객체의 메서드를 사용할 수도 없습니다. View에서는 객체를 다룬다기보다는 데이터를 다룬다고 생각하고 접근하는게 좋습니다. 다만 DTO는 그 변환 과정의 피로도와 코드가 복잡해지는 면이 있고 득보다는 실이 커보이고 과정이 빠른 속도로 진행됨에 따라 지금 단계보다는 Spring을 사용하는 과제때 집중해보는걸 추천드립니다.

Jjungs7 commented on src/main/java/lotto/domain/StatisticsResult.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.Optional;
+
+public class StatisticsResult {
```
DTO, VO 에 관한 개념이 모호했는데 답변해주신 것과 아래 블로그 글까지 모두 보니까 더 명확해진 것 같아요!! 한 가지 더 궁금한 점은 인터넷에서 본 블로그 글 중에 DTO와 VO의 차이는 VO가 read-only라는 점이다라고 언급되어 있는데 제가 생각하기엔 아닌 것 같아요. DTO, VO는 객체 간에 데이터를 주고 받기 위한 구조체이고(즉, get/set 외에는 다른 행위를 하여서는 안됨) DTO는 VO를 포함하는 개념, 하지만 VO는 원시값 하나만을 갖고 있는 가장 기본이 되는 단위로 이해했습니다. 따라서 이번 미션의 [WinningNumbers](https://github.com/next-step/java-lotto-kakao/pull/24/files/86b0b9f97f3d528b768c4960518b18db117c7259#diff-5aae81d8456b88144a182eebb98366de17f0a0557420512b5aee488971c67aa5R5)의 경우 로또 티켓과 보너스볼을 갖고 있으니 DTO이고 나머지 `InsertPrice, LottoNumber, Rate, TotalPrice`는 VO이다가 맞을까요??

---
이번 미션 요구사항 중 **규칙 3: 모든 원시값과 문자열을 포장한다**가 포함되어 있어서 보이는 원시값을 모두 포장하려 했습니다!!

Hue9010 replied:
DTO와 VO가 구성에 있어서는 비슷한 점이 많지만 서로 연관지어 생각하지말고 이름 그대로 Data Transfer Object, Value Object로만 구분하는걸 추천드려요. 계층간 데이터를 전송할 때 사용하기 위한 객체인지, 값으로서 사용 되는 객체인지 말이죠. 단순 getter/setter만 존재해도 데이터 전송을 위해 만든 객체가 아니라면 DTO가 아닙니다. 객체의 형태가 DTO, VO로의 구분을 결정 짓는게 아닙니다.

VO가 read-only인 점은 DTO와 별개로 VO만 고려해서 생각해도 간단합니다. 정확히는 read only라기보다는 불변 객체라는게 맞을 것 같네요. VO가 왜 불변해야하는지는 만원 지폐의 값이 바뀌는 상황을 생각하면 될 것 같네요.(애초에 제대로 된 VO라면 값이 바꼈는데 동일 객체로 본다는게 말이 안되네요.)

참고로 원시값을 포장한다고해서 꼭 VO로 만들 필요는 없습니다. 대표적으로 Integer가 그러한데 원시값을 포장하고 그 외의 기능도 제공을 해주죠. 이번 미션에서 원시값을 포장한 객체를 VO로 만들지 아닐지는 선택사항입니다. 

말씀처럼 xxPrice, Rate, LottoNumber는 VO이고 WinningNumbers는 일반 Object로 보입니다.

Jjungs7 commented on src/main/java/lotto/domain/dto/TotalPrice.java:
```diff
@@ -0,0 +1,22 @@
+package lotto.domain.dto;
+
+public class TotalPrice {
```
규칙 3에 의거해서 유저가 투자한 금액, 유저의 총 당첨 금액이 서로 다른 개념이라 생각돼서 다른 VO로 작성하였지만, 다시 읽어보니 Price라는 큰 틀 안에서 해도 됐을 것 같아요

DTO/VO 에 관하여 자세히 모르는데 다른 분들이 쓰셔서 DTO가 맞나보다 하고 썼던 것 같아요😫 . 제대로 알고 사용하지 않으니까 오히려 혼란을 일으키게 된 것 같네요. 용어를 쓸 땐 확실히 알고 사용 해야겠다는 생각이 들었습니다!!

Jjungs7 commented on src/main/java/lotto/domain/LottoTicketIssuer.java:
```diff
@@ -0,0 +1,22 @@
+    public static LottoTickets issue(NumberPickStrategy strategy, int ticketCount) {
+        return new LottoTickets(IntStream.range(0, ticketCount)
+                .boxed()
+                .map(ignore -> new LottoNumbers(strategy.generateLottoNumbers()))
```
단순 반복문의 경우에는 for, while문으로 하고 list로 어떤 작업을 하고 싶은 경우에는 stream을 사용하는 것이 더 나을까요?
아니면 단순 반복적인 일을 하는 stream의 경우 성능 하락만 잘 유의해서 쓴다면 stream을 사용해도 괜찮을까요?

일반적으로 stream을 사용하는 것이 for, while문보다 깔끔한 코드가 나오는 것 같은데 성능 하락을 방지하기 위해 의식적으로 for, while문을 쓰는 것이 나을 수도 있겠다는 생각이 들어서요!

Hue9010 reviewed (approved):
주말에 빠르게 반영해주셨네요. 👍 
큰 줄기에 변경이 필요한 부분은 아직 반영이 안됐지만 다음 단계의 요구사항을 봤을 때 같이 진행해도 좋을 것 같네요. 이번 단계는 머지할게요. 
질문에 대한 답변 남겼는데 확인하고 다음 단계 진행하면 됩니다! (혹시 제가 놓친 질문이 있다면 말씀 주세요. 😉)
참고로 궁금한 사항은 언제든 질문해주세요.

chanhalee committed:
docs(README) Todo List 작성

Todo List 작성

chanhalee committed:
feat(Budget) 구매 금액에 대한 요구사항 Test 작성 및 구현

구매 금액에 대한 요구사항 Test 작성 및 구현

chanhalee committed:
feat(Ball) 로또 번호에 대한 요구사항 test 작성 및 구현

로또 번호에 대한 요구사항 test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 정수여야 한다.

chanhalee committed:
feat(Ticket) 티켓에 대한 요구사항 Test 작성 및 구현

티켓에 대한 요구사항 Test 작성 및 구현
- 로또 번호는 1 ~ 45 사이의 중복되지 않는 6개의 수로 이루어져 있다

chanhalee committed:
refactor(Budget) validation 메서드 static 속성 제거

validation 메서드 static 속성 제거

chanhalee committed:
feat(WinningNumbers) 티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현

티켓을 당첨 번호와 비교해 등수를 정하는 기능 테스트 작성 및 구현
- 등수에 대한 Enum 클래스 Prize 작성
- Ball -> Number로 클래스 명 변경
- Number Set을 보유하는 Numbers 클래스 생성
- WinningNumbers, TicketNumbers가 Numbers를 상속 받도록 구현

chanhalee committed:
feat(PrizeCalculator) 수익률계산기 테스트 작성 및 구현

수익률계산기 테스트 작성 및 구현

chanhalee committed:
feat(LottoMachine) LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

LottoMachine은 시스템은 구매 금액에 상응하는 개수의 로또 번호를 만들어준다

chanhalee committed:
feat(LottoMachine) LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다

LottoMachine은 당첨된 로또에 대한 통계와 수익률을 집계한다
